rpcs3/rpcs3qt/skylander_dialog.cpp

```diff
@@ -642,273 +531,307 @@ u16 skylander_dialog::skylander_crc16(u16 init_value, const u8* buffer, u32 size
-	for (u32 index = 1; index < 0x10; index++)
-	{
-		for (u32 subdex = 0; subdex < (0x30 / sizeof(u64)); subdex++)
-		{
-			if (reinterpret_cast<const u64&>(g_skylander.sky_dump[(index * 0x40) + (subdex * sizeof(u64))]) != 0)
-			{
-				return true;
-			}
-		}
-	}
-	return false;
+	return file_path;
 }
 
-void skylander_dialog::new_skylander()
+skylander_dialog::skylander_dialog(QWidget* parent)
+    : QDialog(parent)
 {
-	const QString file_path = QFileDialog::getSaveFileName(this, tr("Create Skylander File"), cur_sky_file_path, tr("Skylander Object (*.sky);;"));
-	if (file_path.isEmpty())
-		return;
+	setWindowTitle(tr("Skylanders Manager"));
+	setObjectName("skylanders_manager");
+	setAttribute(Qt::WA_DeleteOnClose);
+	setMinimumSize(QSize(700, 200));
 
-	if (g_skylander.sky_file)
-		g_skylander.sky_file.close();
+	QVBoxLayout* vbox_panel = new QVBoxLayout();
 
-	g_skylander.sky_file.open(file_path.toStdString(), fs::read + fs::write + fs::create);
-	if (!g_skylander.sky_file)
-		return;
+	auto add_line = [&](QVBoxLayout* vbox)
+	{
+		QFrame* line = new QFrame();
+		line->setFrameShape(QFrame::HLine);
+		line->setFrameShadow(QFrame::Sunken);
+		vbox->addWidget(line);
+	};
 
-	cur_sky_file_path = file_path;
+	QGroupBox* group_skylanders = new QGroupBox(tr("Active Portal Skylanders:"));
+	QVBoxLayout* vbox_group     = new QVBoxLayout();
 
+	for (auto i = 0; i < UI_SKY_NUM; i++)
 	{
-		std::lock_guard lock(g_skylander.sky_mutex);
-		memset(g_skylander.sky_dump, 0, 0x40 * 0x10);
+		QHBoxLayout* hbox_skylander = new QHBoxLayout();
+		QLabel* label_skyname       = new QLabel(QString(tr("Skylander %1")).arg(i + 1));
+		edit_skylanders[i]          = new QLineEdit();
+		edit_skylanders[i]->setEnabled(false);
 
-		// Set the block permissions
-		reinterpret_cast<le_t<u32>&>(g_skylander.sky_dump[0x36]) = 0x690F0F0F;
-		for (u32 index = 1; index < 0x10; index++)
-		{
-			reinterpret_cast<le_t<u32>&>(g_skylander.sky_dump[(index * 0x40) + 0x36]) = 0x69080F7F;
-		}
+		QPushButton* clear_btn  = new QPushButton(tr("Clear"));
+		QPushButton* create_btn = new QPushButton(tr("Create"));
+		QPushButton* load_btn   = new QPushButton(tr("Load"));
+
+		connect(clear_btn, &QAbstractButton::clicked, this, [=, this]() { this->clear_skylander(i); });
+		connect(create_btn, &QAbstractButton::clicked, this, [=, this]() { this->create_skylander(i); });
+		connect(load_btn, &QAbstractButton::clicked, this, [=, this]() { this->load_skylander(i); });
 
-		reinterpret_cast<le_t<u16>&>(g_skylander.sky_dump[0x1E]) = skylander_crc16(0xFFFF, g_skylander.sky_dump, 0x1E);
+		hbox_skylander->addWidget(label_skyname);
+		hbox_skylander->addWidget(edit_skylanders[i]);
+		hbox_skylander->addWidget(clear_btn);
+		hbox_skylander->addWidget(create_btn);
+		hbox_skylander->addWidget(load_btn);
 
-		// On a new skylander everything is 0'd and no crc apart from the first one is set
+		vbox_group->addLayout(hbox_skylander);
 
-		g_skylander.sky_reload = true;
+		if (i != (UI_SKY_NUM - 1))
```

Comment:
Just wondering what this is

Reply:
It's so it adds a line after each Skylander UI line but not the last?

Reply:
And if you're wondering why it's governed by UI_SKY_NUM it's because in theory the portal can support up to 8 figures at the same time but afaik no game goes beyond 3 or 4 figures(and UI would be too busy with 8 different skylanders).

In case I'm proven wrong I just have to change var to change the number of selectable skylanders.

Reply:
if (i != 0) at the **start** of the scope does the same thing but better looking.

Reply:
I want the line after the layout for the skylander information/buttons.
(i != 0) would add a line before all but the first when I want to add a line after each but the last?